python/lsst/daf/butler/registry/queries/exprParser/normalForm.py

```diff
@@ -0,0 +1,1137 @@
+        the special structure guaranteed by the normal forms.
+
+        Parameters
+        ----------
+        visitor : `NormalFormVisitor`
+            Visitor object to apply.
+
+        Returns
+        -------
+        result
+            Return value from calling ``visitor.visitOuter`` after visiting
+            all other nodes.
+        """
+        visitedOuterBranches: List[_U] = []
+        for nodeInnerBranches in self._nodes:
+            visitedInnerBranches = [visitor.visitBranch(node) for node in nodeInnerBranches]
+            visitedOuterBranches.append(visitor.visitInner(visitedInnerBranches, self.form))
+        return visitor.visitOuter(visitedOuterBranches, self.form)
+
+    def toTree(self) -> Node:
+        """Convert ``self`` back into an equivalent tree, preserving the
+        form of the boolean expression but representing it in memory as
+        a tree.
+
+        Returns
+        -------
+        tree : `Node`
+            Root of the tree equivalen to ``self``.
+        """
+        visitor = TreeReconstructionVisitor()
+        return self.visit(visitor)
+
+
+class PrecedenceTier(enum.Enum):
+    """An enumeration used to track operator precedence.
+
+    This enum is currently used only to inject parentheses into boolean
+    logic that has been manipulated into a new form, and because those
+    parentheses are only used for stringification, the goal here is human
+    readability, not precision.
+
+    Lower enum values represent tighter binding, but code deciding whether to
+    inject parentheses should always call `needsParens` instead of comparing
+    values directly.
+    """
+
+    TOKEN = 0
+    """Precedence tier for literals, identifiers, and expressions already in
+    parentheses.
+    """
+
+    UNARY = 1
+    """Precedence tier for unary operators, which always bind more tightly
+    than any binary operator.
+    """
+
+    VALUE_BINARY_OP = 2
+    """Precedence tier for binary operators that return non-boolean values.
+    """
+
+    COMPARISON = 3
+    """Precendence tier for binary comparison operators that accept non-boolean
+    values and return boolean values.
+    """
+
+    AND = 4
+    """Precendence tier for logical AND.
+    """
+
+    OR = 5
+    """Precedence tier for logical OR.
+    """
+
+    @classmethod
+    def needsParens(cls, outer: PrecedenceTier, inner: PrecedenceTier) -> bool:
+        """Test whether parentheses should be added around an operand.
+
+        Parameters
+        ----------
+        outer : `PrecedenceTier`
+            Precedence tier for the operation.
+        inner : `PrecedenceTier`
+            Precedence tier for the operand.
+
+        Returns
+        -------
+        needed : `bool`
+            If `True`, parentheses should be added.
+
+        Notes
+        -----
+        This method special cases logical binary operators for readability,
+        adding parentheses around ANDs embedded in ORs, and avoiding them in
```

Comment:
No, this is what I intended - I'm saying that I add parens around ANDs even though they're not formally necessary, because (I think) it improves readability.